src/devices/src/virtio/device.rs

```diff
@@ -10,6 +10,14 @@ use std::sync::{atomic::AtomicUsize, Arc};
 use super::{ActivateResult, Queue};
 use crate::virtio::AsAny;
 use utils::eventfd::EventFd;
+use vm_memory::GuestMemoryMmap;
+
+/// Enum that indicates if a VirtioDevice is inactive or has been activated
+/// and memory attached to it.
+pub enum DeviceState {
+    Inactive,
+    Activated(GuestMemoryMmap),
```

Comment:
Having the `GuestMemoryMmap` here doesn't seem very flexible. If we want to have another state that also just applies to activated devices, we would need to convert this to a tuple. The code becomes harder to read then. As we are thinking about implementing reset on devices, the time where extending Activated is not that far because on reset we need to release some other resources.

Reply:
Each state variant can have its own data (including overlapping data) so we wouldn't use tuples, just another state with or without `mem`.

Regarding `reset` support, the changes in this PR should help in that scenario.

Adding `mem` to `enum DeviceState` was just a solution that felt natural to me, but which I'm not hell-bent on doing.

The problem which this PR is trying to solve is allowing devices to be created without `guest memory`. Memory will be added during device activation. We need *some way* to allow this plugging of memory after creation.

One solution is `Option<Memory>`, but
```rust
struct SomeDevice {
    ...
    is_active: bool,
    mem: Option<Memory>,
    ...
}
```
is C-style code where you have an invisible error-prone connection between the two fields.

In Rust you can enforce that connection at compile time using an enum like the one in the PR.

```rust
enum DevState {
    Inactive,
    Active(Memory),
};
struct SomeDevice {
    ...
    state: DevState,
    ...
}
```

I'm open to alternatives.

Reply:
I was discussing about this with @dhrgit at some point. Hopefully the following lines are a good summary of our discussion.

Devices should always be initialized after the memory is initialized because this is also how it works in a real world scenario. When you power on a machine, the memory is initialized, then the devices use the memory to initialize themselves. This would be mapped in code with the device constructor always receiving a reference to the memory. This way you also break the dependency between `mem` and `is_active`. The device cannot exist without the memory, so the `mem` field will always be `Memory` and not `Option<Memory>`. The activation can happen later in the logic of the device and does not have any impact on the memory.

Reply:
There were also discussions with both @dhrgit and @alexandruag where we were comparing a device with a physical device which is first built (like in a factory in China 😆 ) and at some later point is, like you say initialized when it's plugged in a machine. The device can surely exist without `memory` or anything else that is outside of it.

Where I am trying to get with this change and some following ones pending on this, is a place where a device can be _built_ during it's configuration phase as opposed to being built during boot. Booting a microVM then becomes a matter of plugging all things together and hitting the switch.

This model allows user-errors (like invalid resources or permissions) to be reported as part of the configuration step rather than later at attempted boot.
It also removes the time window between configuration and boot where changes to the host system (creating Taps, changing permissions, etc) could cause microVM boot failures (race condition between validation and commitment of resources).

Such a model should also decreases code complexity and should provide better efficiency to the process of configuring and starting a microVM (no more moving configurations around in memory and validating resources both at config time and boot-time).

Reply:
I am not necessarily against that model. But you can have both of them by introducing a restriction at the API level that the memory needs to be configured before you configure the devices.

Indeed, not all devices need memory, but network, block and vsock do need it and these are the ones we are trying to model now.

Reply:
That is a valid alternative!

Though, I personally think that such a limitation at the API level increases UX complexity as the user doesn't naturally understand such an ordering requirement (and it gets even more confusing when supporting default values for memory like we do), while the current proposal has no disadvantages AFAICT.

If you like the API-level restriction better we can set up a meeting with the maintainers, get busy with the whiteboard and commit to an option. Let me know.

Reply:
Yap. I would prefer a meeting because this approach is somewhat going in a very different direction compared to the rust-vmm approach.

Reply:
This is the interface that is currently merged in rust-vmm: https://github.com/rust-vmm/vm-virtio/blob/master/src/device.rs

It will probably suffer some changes though once we add more functionality to the crate.

Reply:
> this approach is somewhat going in a very different direction compared to the rust-vmm approach

How so? The rust-vmm device interface _already_ has memory passed to the device during activation (exactly what this PR proposes).

Reply:
That is a temporary state because the code was very similar to what we had in Firecracker before. You can take a look at what adjustment are in the pipeline: https://github.com/rust-vmm/vm-virtio/pull/10 and https://github.com/rust-vmm/vm-virtio/pull/13.

When we discussed about unifying the activation path I though we will apply the same mechanism from vsock without introducing other changes, so I am a bit unprepared for a design discussion and I would need more time to think about this. I personally prefer the vsock design for now. This can be improved in the future, but it probably needs more investment. I find it more clear for `device_activated` to be a bool and the memory to be directly accessible from a field (it doesn't need to be field of the VirtioDevice, we can have it in a struct if we find something that makes sense) and not from an enum variant. I also strongly believe that virtio devices should not be initialized without a reference to the memory.

It looks like this PR is in preparation for a bigger change that is yet to come, so the full picture is needed to understand if this change is indeed need in its current form. Of course, we wouldn't need to block on the next PR availability, but a few paragraphs on the overall design would be helpful.

Of course all my rambling can come off as a personal preference, so I think it is better to leave it at the submitter choice because otherwise is going to be a too long discussion on something that can be updated later if we realize that's not the right direction.

Reply:
> You can take a look at what adjustment are in the pipeline: rust-vmm/vm-virtio#10 and rust-vmm/vm-virtio#13.

Both of these are compatible with the model proposed in this PR. In https://github.com/rust-vmm/vm-virtio/pull/10 you can even see that the `VirtioDevice` trait defines activation as:

```rust
    /// Associated guest memory
    type M: GuestMemory;

    ...

    /// Activates this device for real usage.
    /// The ownership of the VirtioDeviceConfig object moves into the VirtioDevice object if
    /// activate succeeds, otherwise it should return the ownership back.
    fn activate(&mut self, config: VirtioDeviceConfig<M>) -> ActivateResult<VirtioDeviceConfig<M>>;
```

So the model is to pass virtio device configuration (including memory - the `M` in `VirtioDeviceConfig`) to the device during its activation.

This is a sane model. Most of the configuration is not available during device creation, some of it can be dynamic (like in the case of PCI). 

-----------------

> I also strongly believe that virtio devices should not be initialized without a reference to the memory.

Can you provide some supporting data/theories for that?

-----------------

> It looks like this PR is in preparation for a bigger change that is yet to come, so the full picture is needed to understand if this change is indeed need in its current form.

That is right, it is required, although not necessarily in its current form, for the following story: https://github.com/firecracker-microvm/firecracker/issues/1702 https://github.com/firecracker-microvm/firecracker/issues/1708 https://github.com/firecracker-microvm/firecracker/issues/1709 and finally https://github.com/firecracker-microvm/firecracker/issues/1713

The end result is reaching a state where our components are loose enough so we can easily swap them for [rust-vmm](https://github.com/rust-vmm) ones.

Reply:
I looked again at the code, my bad. I thought that `DeviceState` leaks in the interface and that would've been a problem. As long as `DeviceState` is not leaking in the interface, we can easily replace this abstraction with something else in the future (for example when consuming virito devices from rust-vmm).

Nevertheless, I don't think that `DeviceState` is the right abstraction as it complicates the code (with the match in process and deciding whether a device is activated or not). I still don't understand why not use the vsock design, but again it is too long of a conversation for a few lines of code.

Regarding the virtio devices requiring memory to be created, we are using static devices, and no PCI. So for us these resources can be created after the memory is initialized, which would simplify things because you don't need an Option<Memory> or the memory to be obfuscated behind the DeviceState which again would make access to the memory very simple instead of a match.